blockprod/src/detail/block_solver.rs

```diff
@@ -0,0 +1,726 @@
+// Copyright (c) 2021-2024 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::{BTreeMap, BTreeSet},
+    sync::Arc,
+};
+
+use chainstate_types::pos_randomness::PoSRandomness;
+use crypto::{
+    key::PrivateKey,
+    vrf::{VRFPrivateKey, VRFReturn},
+};
+use serialization::Encode;
+use tokio::sync::{oneshot, watch};
+
+use ::utils::{atomics::RelaxedAtomicBool, once_destructor::OnceDestructor};
+use chainstate::{chainstate_interface::ChainstateInterface, GenBlockIndex};
+use common::{
+    chain::{
+        block::{
+            block_body::BlockBody,
+            consensus_data::PoSData,
+            signed_block_header::{
+                BlockHeaderSignature, BlockHeaderSignatureData, SignedBlockHeader,
+            },
+            timestamp::BlockTimestamp,
+            BlockHeader, BlockReward, ConsensusData,
+        },
+        config::EpochIndex,
+        Block, ChainConfig, GenBlock, PoSChainConfig, PoWStatus, PoolId, RequiredConsensus,
+    },
+    primitives::{BlockHeight, Compact, Id},
+    Uint256,
+};
+use consensus::{
+    calculate_target_required_from_block_index, compact_target_to_target,
+    generate_pos_consensus_data_and_reward, generate_pow_consensus_data_and_reward,
+    generate_reward_ignore_consensus, mine, stake, ConsensusCreationError, ConsensusPoSError,
+    ConsensusPoWError, GenerateBlockInputData, MiningResult, PoSGenerateBlockInputData,
+    PoSPartialConsensusData, PoSSlotInfo, PoSSlotInfoCmpByParentTS, PoWGenerateBlockInputData,
+    StakeResult,
+};
+
+use crate::BlockProductionError;
+
+use super::{
+    utils::{
+        get_existing_gen_block_index, get_pool_balances_at_heights, get_sealed_epoch_randomness,
+        make_ancestor_getter, timestamp_add_secs,
+    },
+    BlockProduction, TxData,
+};
+
+impl BlockProduction {
```

Comment:
I'm not a big fan of having a single struct's impls in multiple files... this indicates that some restructuring might be necessary with better encapsulation and isolated responsibilities.